perf/src/deduper.rs

```diff
@@ -241,42 +253,19 @@ mod tests {
         ));
     }
 
-    #[test_case([0xf9; 32],  3_199_997, 101_192,  51_414,  66, 101_121)]
-    #[test_case([0xdc; 32],  3_200_003, 101_192,  51_414,  60, 101_092)]
-    #[test_case([0xa5; 32],  6_399_971, 202_384, 102_828, 125, 202_178)]
-    #[test_case([0xdb; 32],  6_400_013, 202_386, 102_828, 135, 202_235)]
-    #[test_case([0xcd; 32], 12_799_987, 404_771, 205_655, 285, 404_410)]
-    #[test_case([0xc3; 32], 12_800_009, 404_771, 205_656, 293, 404_397)]
-    fn test_dedup_seeded(
```

Comment:
This change is reducing test coverage by a lot.
Also I think the existing `test_case`s provides a clear explicit demonstration of how the deduper is expected to work.
There is also in fact some value in these hard-coded values because it helps to catch unexpected changes in behavior of the deduper after an upgrade or on a certain machine, etc.

Does updating the values inside `test_case`s unblock you for https://github.com/anza-xyz/agave/pull/3254?
If so, I think the best is to just update these and keep the test coverage as before. 

If someone still complained that the test failed on their machine, we can think of a different solution which does not reduce coverage.

Reply:
What cases are not covered? We're not running a bunch of random loops so there's less data going through them, but we're instead targetting specific cases for the tests.

> There is also in fact some value in these hard-coded values because it helps to catch unexpected changes in behavior of the deduper after an upgrade or on a certain machine, etc.

Sure there's value in testing the actual deduping, but there's not value in testing the actual hash values.

We could generated random (seeded) u64s for our hashes, and do similar large loops with varied parameters and make sure the values do not change, wdyt?
But leaving the tests as is, i.e. relying on consistent hashes for a library that doesn't guarantee that, is not a good idea imo.

Reply:
> but there's not value in testing the actual hash values

I disagree with that. If after an update to ahash or on a certain machine etc we get too many false positive, the hard-coded values helps us to notice that.

> relying on consistent hashes for a library that doesn't guarantee that, is not a good idea

We need to guarantee that deduper has sane behavior.
If an update to the library breaks that guarantee, then we need some way to notice that and perhaps avoid the library update, or use a different hasher, etc. 

I still think if updating these values unblocks you for https://github.com/anza-xyz/agave/pull/3254 then the best is to just do that.

Reply:
Lets say ahash has a malicious update which always returns 42 no matter what the input to the hasher is.
Obviously that breaks the deduper, and the existing test catches that and prevents us from updating to that malicious version of the library.

I am making an extreme example to make the point across but I am sure you can think of more subtle ways a library can  rug pull us, in which case such hard-coded tests will be useful.

Reply:
and I'm saying then let's add a separate test that tests JUST collision rate, which is the concern.

The deduper tests are fragile because of the current structure relying on exact hash value consistency. We know that isn't there so we should not rely on it.

Much more than deduper is at risk if a hash had a malicious update like that.
A distinct test makes it very obvious if such a regression occurs.
The current tests just makes people update hard coded constants without any understanding of what is actually wrong

Reply:
To be clear I will add these tests. I'm not going to merge this PR as is.

But I want to just get buy-in from you before I go out and write them

Reply:
I think you are just repeating previous statements and if I want to argue I have to repeat my previous responses.

>  The current tests just makes people update hard coded constants without any understanding of what is actually wrong

If the reviewer has enough context on why that test is there, he can capture problematic updates.

Reply:
I pushed the test(s) with removed asserts on exact pop count and dup count since those are the problem with inconsistent hashing. The tests simply check the false postive rate.

This, I hope, resolves your concerns about hashing quality regression, and I think this is reasonable.

Reply:
> removed asserts on exact pop count and dup count

This still means reducing test coverage and I frankly do not like that.

I am updating ahash to 0.8.11 in https://github.com/anza-xyz/agave/pull/3267.
If all is green there I will merge that and you can rebase.

That will unblock your work without reducing test-coverage so I can sleep well at night.

Reply:
I'm not trying to simply unblock my work. That's just how I found the issue. I'm trying to improve tests because from my perspective they are badly designed.

Reply:
My point is that I am not convinced this change is an improvement to the test.

We have to guarantee that the deduper still behaves reasonably even if the dependencies don't provide certain guarantees. The test was intentionally written like that to assert that.